kroxylicious-schema-validation/src/main/java/io/kroxylicious/proxy/filter/schema/ProduceValidationFilter.java

```diff
@@ -0,0 +1,122 @@
+ */
+package io.kroxylicious.proxy.filter.schema;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.kafka.common.message.ProduceRequestData;
+import org.apache.kafka.common.message.ProduceResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.Errors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.ProduceRequestFilter;
+import io.kroxylicious.proxy.filter.ProduceResponseFilter;
+import io.kroxylicious.proxy.filter.schema.validation.ProduceRequestValidator;
+import io.kroxylicious.proxy.filter.schema.validation.RecordValidationFailure;
+import io.kroxylicious.proxy.filter.schema.validation.TopicValidationResult;
+import io.kroxylicious.proxy.filter.schema.validation.ValidationResult;
+
+import static java.util.stream.Collectors.joining;
+
+public class ProduceValidationFilter
+        implements ProduceRequestFilter, ProduceResponseFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProduceValidationFilter.class);
+    private final ProduceRequestValidator validator;
+    private final Map<Integer, ValidationResult> correlatedResults = new HashMap<>();
+
+    public ProduceValidationFilter(ProduceRequestValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void onProduceRequest(RequestHeaderData header, ProduceRequestData request, KrpcFilterContext context) {
+        ValidationResult result = validator.validateRequest(request);
+        if (result.allTopicPartitionsInvalid()) {
+            LOGGER.warn("all topic-partitions for request contained invalid data: {}", result);
+            ProduceResponseData response = invalidateAllTopicPartitionProduceRequests(result);
+            context.forwardResponse(response);
+        }
+        else if (result.anyTopicPartitionInvalid()) {
+            LOGGER.warn("some topic-partitions for request contained invalid data: {}", result);
+            request.topicData().removeIf(topicProduceData -> result.allPartitionsInvalid(topicProduceData.name()));
+            for (ProduceRequestData.TopicProduceData topicDatum : request.topicData()) {
+                topicDatum.partitionData().removeIf(partitionProduceData -> !result.valid(topicDatum.name(), partitionProduceData.index()));
+            }
+            correlatedResults.put(header.correlationId(), result);
+            context.forwardRequest(request);
+        }
+        else {
+            context.forwardRequest(request);
+        }
+    }
+
+    private static ProduceResponseData invalidateAllTopicPartitionProduceRequests(ValidationResult request) {
+        ProduceResponseData response = new ProduceResponseData();
```

Comment:
Did you see the `org.apache.kafka.common.requests.AbstractRequest#getErrorResponse(int, java.lang.Throwable)` factories?  Not sure it would actually help in this case, but I thought I'd mention it as it might have utility in other filters that short circuit.